src/main/python/ttconv/isd.py

```diff
@@ -0,0 +1,346 @@
+
+  def iter_regions(self) -> typing.Iterator[ISD.Region]:
+    '''Returns an iterator over regions.'''
+    return self._regions.values()
+
+  # creating an ISD from a document
+
+  @staticmethod
+  def significant_times(doc: model.Document) -> typing.Set[Fraction]:
+    '''Returns the temporal offsets at which the document `doc` changes
+    '''
+
+    def make_absolute(
+        begin_offset: typing.Optional[Fraction],
+        end_offset: typing.Optional[Fraction],
+        parent_begin: typing.Optional[Fraction],
+        parent_end: typing.Optional[Fraction]
+      ) -> typing.Tuple[Fraction, Fraction]:
+      
+      begin_time = (parent_begin or Fraction(0)) + (begin_offset or Fraction(0))
+
+      end_time = (parent_begin or Fraction(0)) + end_offset if end_offset is not None else None
+
+      if end_time is None:
+
+        end_time = parent_end
+
+      elif parent_end is not None:
+
+        end_time = min(end_time, parent_end)
+
+      return (begin_time, end_time)
+
+    def sig_times(element: model.ContentElement, parent_begin: Fraction, parent_end: typing.Optional[Fraction]):
+
+      # add signficant times for the element
+
+      begin_time, end_time = make_absolute(element.get_begin(), element.get_end(), parent_begin, parent_end)
+
+      s_times.add(begin_time)
+
+      if end_time is not None:
+        s_times.add(end_time)
+
+      # add signficant times for any animation step 
+
+      for anim_step in element.iter_animation_steps():
+        anim_begin_time, anim_end_time = make_absolute(anim_step.begin, anim_step.end, parent_begin, parent_end)
+
+        s_times.add(anim_begin_time)
+
+        if anim_end_time is not None:
+          s_times.add(anim_end_time)
+
+      # add signficant times for the children of the element 
+
+      for child_element in iter(element):
+        sig_times(child_element, begin_time, end_time)
+
+    s_times = set()
+
+    # add significant times for regions
+
+    for region in doc.iter_regions():
+      sig_times(region, 0, None)
+
+    # add significant times for body and its descendents
+
+    sig_times(doc.get_body(), 0, None) 
+
+    return s_times
+
+  @staticmethod
+  def from_model(doc: model.Document, offset: Fraction) -> typing.Optional[ISD]:
+    '''Creates an ISD from a snapshot of a Document `doc` at a given time offset `offset`
+    '''
+    isd = ISD(doc)
+
+    for region in doc.iter_regions():
+      root_region = ISD.process_element(isd, offset, region, None, None, region)
+
+      if root_region is not None:
+        isd.put_region(root_region)
+
+    return isd
+
+
+  @staticmethod
+  def compute_style(_style_property: styles.StyleProperty, value: typing.Any) -> typing.Any:
+    return value
+
+  @staticmethod
+  def process_element(isd: ISD,
+                      offset: Fraction,
+                      selected_region: model.Region,
+                      inherited_region: typing.Optional[model.Region],
+                      parent: model.ContentElement,
```

Comment:
`parent` should be Optional since it is `None` here: https://github.com/sandflow/ttconv/pull/109/files#diff-612df484ef6f59842b0bfc09650955d79933017b5884740294e7b4c029979a01R159